### PR TITLE
Fix for issue #981

### LINKED ISSUE #981: Starting a standby cache problems
## Description

Behavior description:
Start-cache command with the standby flag shall accept only the cache-device, the cache-id and the cache-line-size parameters. All of the parameters shall be mandatory.
Start-cache command flags for:
•	loading a cache instance,
•	forcing cache initialization,
•	starting a standby cache instance (standby flag)
shall be mutually exclusive.

## Expected Behavior

Invalid operations should be blocked.

## Actual Behavior

CAS accepts disallowed parameters with the standby flag and allows for starting a standby cache instance without the required cache-id and cache-line-size parameters.
Combination of standby and load flags works as loading cache.

## Steps to Reproduce

1. `casadm --start-cache --cache-device <DEVICE> --failover-standby`
2. `casadm --start-cache --cache-device <DEVICE> --failover-standby --cache-id 1 --cache-line-size 4 --cache-mode wb`
3. `casadm --start-cache --cache-device <DEVICE> --failover-standby --load` results in loading cache instead of
```
Use of 'failover-standby' and 'load' simultaneously is forbidden.
```
(Works ok for 'force').

## Your Environment

* OpenCAS version (commit hash or tag): 21.12.0.0579.master

### PR DESCRIPTION
**Pull request by kmajzero**
Fixes #981

### CODE REVIEW

**Review by Deixx — CHANGES_REQUESTED** [review]
It doesn't fix the issue where you can pass additional arguments (cache-mode) for failover-standby, does it?

**Comment by kmajzero** [comment]
> It doesn't fix the issue where you can pass additional arguments (cache-mode) for failover-standby, does it?

yes, fixed :)

File: casadm/cas_main.c [hunk 1]
```diff
@@ -70,7 +70,7 @@ static struct command_args command_args_values = {
 		.stats_filters = STATS_FILTER_DEFAULT,
 		.output_format = OUTPUT_FORMAT_DEFAULT,
 		.io_class_id = OCF_IO_CLASS_INVALID,
-		.line_size = ocf_cache_line_size_default,
+		.line_size = ocf_cache_line_size_none, //ocf_cache_line_size_default,
```
Comment (rafalste) [inline casadm/cas_main.c:73]: Is there any purpose for leaving commented default cache line size?

File: casadm/cas_main.c [hunk 2]
```diff
@@ -197,9 +197,21 @@ int start_cache_command_handle_option(char *opt, const char **arg)
...
 		command_args_values.cache_id = atoi(arg[0]);
 	} else if (!strcmp(opt, "load")) {
-		command_args_values.state = CACHE_INIT_LOAD;
+		if (command_args_values.state == CACHE_INIT_STANDBY) {
+			cas_printf(LOG_ERR, "Use of 'failover-standby' and 'load' simultaneously is forbidden.\n");
+			return FAILURE;
+		}
+		else {
+			command_args_values.state = CACHE_INIT_LOAD;
+		}
```
Comment (rafalste) [inline casadm/cas_main.c:206]: No need for _else_ here.

File: casadm/cas_main.c [hunk 3]
```diff
@@ -197,9 +197,21 @@ int start_cache_command_handle_option(char *opt, const char **arg)
...
 	} else if (!strcmp(opt, "failover-standby")) {
-		command_args_values.state = CACHE_INIT_STANDBY;
+		if (command_args_values.state == CACHE_INIT_LOAD) {
+			cas_printf(LOG_ERR, "Use of 'failover-standby' and 'load' simultaneously is forbidden.\n");
+			return FAILURE;
+		}
+		else {
+			command_args_values.state = CACHE_INIT_STANDBY;
+		}
```
Comment (rafalste) [inline casadm/cas_main.c:214]: here also

File: casadm/cas_main.c [hunk 4]
```diff
@@ -334,9 +346,24 @@ int handle_start()
...
+	}
+
+	if (command_args_values.line_size == ocf_cache_line_size_none) {
+		if (command_args_values.state == CACHE_INIT_STANDBY) {
+			cas_printf(LOG_ERR, "Use of 'failover-standby' without 'cache-line-size' explicitly specified is forbidden.\n");
+			return FAILURE;
+		} else {
+			command_args_values.line_size = ocf_cache_line_size_default;
+		}
```
Comment (rafalste) [inline casadm/cas_main.c:366]: and here :)

File: casadm/cas_main.c (hunk 1 above)
Comment (kmajzero) [reply casadm/cas_main.c:73]: comments are already removed :)

**Review by rafalste — APPROVED** [review]

File: casadm/cas_main.c [hunk 5]
```diff
@@ -334,9 +344,33 @@ int handle_start()
...
 
-	if (command_args_values.state == CACHE_INIT_STANDBY && command_args_values.force) {
-		cas_printf(LOG_ERR, "Use of 'failover-standby' and 'force' simultaneously is forbidden.\n");
-		return FAILURE;
+	if (command_args_values.state == CACHE_INIT_STANDBY) {
+		if (command_args_values.force) {
+			cas_printf(LOG_ERR, "Use of 'failover-standby' and 'force' simultaneously is forbidden.\n");
+			return FAILURE;
+		}
+		if (command_args_values.cache_id == 0) {
```
Comment (Deixx) [inline casadm/cas_main.c:352]: Maybe compare with OCF_CACHE_ID_INVALID instead of 0?

**Review by Deixx — CHANGES_REQUESTED** [review]

File: casadm/cas_main.c [hunk 6]
```diff
@@ -334,9 +344,33 @@ int handle_start()
...
+			cas_printf(LOG_ERR, "Use of 'failover-standby' without 'cache-id' explicitly specified is forbidden.\n");
```
Comment (robertbaldyga) [inline casadm/cas_main.c:353]: What we want to say to user is rather something like "'failover-standby' requires specifying 'cache-id'". 
That would be a meaningful hint.

**Review by robertbaldyga — APPROVED** [review]

**Review by Deixx — APPROVED** [review]

**Pull request merged**